### PR TITLE
SS-29663: [MOL V3000 writer] Writes S-Groups

### PR DESCRIPTION
**Pull request by apar03**
- Adds MOL V3000 writer to Ketcher, which constructs the MOL V3000 representation from the underlying [`struct`](https://github.com/schrodinger/ketcher-schrodinger/blob/master/src/script/chem/struct/index.js) data structure.
- This PR addresses writing S-Groups in V3000 format, as described in the [CTFile Representation Documentation, Chapter 10: The Extended Connection Table (V3000)](https://drive.google.com/file/d/0Bx3dsPc7eyZKdXlHTktQTFJUMHc/view)
- Tested with data S-Groups and repeating chains etc. 
- Please note that the S-Group action in Ketcher (i.e., creating S-Groups by using the tool icon) is currently faulty (https://jira.schrodinger.com/browse/SS-29662). However, the V3000 reading (parsing) logic is correct. So, for testing, 
1. sketch S-Groups in an external sketcher (say Marvin), and export the V3000 representation
2. import this representation to Ketcher,
3. now export the representation from Ketcher in V3000 format
4. import this representation back to the external sketcher to verify the structure is exactly the same

Primary - @mycrobe

### CODE REVIEW

File: src/script/chem/molfile/v3000.js [hunk 1]
```diff
@@ -162,6 +162,8 @@ function v3000parseSGroup(ctab, ctabLines, sgroups, atomMap, shift) { // eslint-
 			sg.data.connectivity = props['CONNECT'][0].toLowerCase();
 		if (props['FIELDDISP'])
 			sGroup.applyDataSGroupInfo(sg, stripQuotes(props['FIELDDISP'][0]));
+		if (props['FIELDINFO'])
+			sg.data.units = props['FIELDINFO'][0];
```
Comment (apar03) [inline src/script/chem/molfile/v3000.js:166]: noticed from Marvin that the units for a Data S-Group are written out as FIELDINFO=<units>

File: src/script/chem/molfile/molfile.js [hunk 2]
```diff
@@ -286,9 +286,135 @@ Molfile.prototype.writeCTab3000 = function (molecule) { // eslint-disable-line m
...
 
+Molfile.prototype.writeSGroupBlock3000 = function (molecule) {
+	if (molecule.sgroups.size === 0) {
+		return;
+	}
+	this.writeCR('M  V30 BEGIN SGROUP');
+	var sgmap = {};
+	var cnt = 1;
+	var sgmapback = {};
+	var sgorder = molecule.sGroupForest.getSGroupsBFS();
```
Comment (mycrobe) [inline src/script/chem/molfile/molfile.js]: could you make these variable names a little more descriptive and camelCase?

File: src/script/chem/molfile/molfile.js [hunk 3]
```diff
@@ -286,9 +286,135 @@ Molfile.prototype.writeCTab3000 = function (molecule) { // eslint-disable-line m
...
+	sgorder.forEach((id) => {
+		sgmapback[cnt] = id;
+		sgmap[id] = cnt++;
```
Comment (mycrobe) [inline src/script/chem/molfile/molfile.js]: I'm not sure how others feel here, but I'd prefer to have ++increments inside a for loop, or use the second `index` arg of the forEach to keep track of index.

File: src/script/chem/molfile/molfile.js [hunk 4]
```diff
@@ -286,9 +286,135 @@ Molfile.prototype.writeCTab3000 = function (molecule) { // eslint-disable-line m
...
+	});
+	for (var q = 1; q < cnt; ++q) { // each group on its own
```
Comment (mycrobe) [inline src/script/chem/molfile/molfile.js]: Isn't `cnt` going to be 1 greater than the size of `sgmapback` and `sgmap`?

File: src/script/chem/molfile/molfile.js (hunk 2 above)
Comment (mycrobe) [inline src/script/chem/molfile/molfile.js]: What is the data structure of `sgorder`? I assume it is an array, but what does it contain?

File: src/script/chem/molfile/molfile.js [hunk 5]
```diff
@@ -286,9 +286,135 @@ Molfile.prototype.writeCTab3000 = function (molecule) { // eslint-disable-line m
...
+		var id = sgmapback[q];
+		var sgroup = molecule.sgroups.get(id);
+		let sgDetails = `${this.getMol3000Prefix()}${id + 1} ${sgroup.type} 0`;
+		if (sgroup.atoms.length) {
+			// atoms that define the Sgroup
+			// ATOMS=(natoms aidx1 aidx2 ..)
+			sgDetails += ` ATOMS=(${sgroup.atoms.length} ${sgroup.atoms.map(atom => atom + 1).join(' ')})`;
```
Comment (mycrobe) [inline src/script/chem/molfile/molfile.js:311]: I assume that mol files are 1-indexed and the ketcher internal data structure is zero indexed.

Thought/Idea: It *might* be worth having some named util functions to do this? `toCtabIdx(ketcherIdx)`, and the inverse. I realize that the computation is incredibly trivial but there are *a lot* of `someVariable + 1` operations in this block of code... 

With named util functions you could express, e.g. `sgroup.atoms.map(toCtabIdx)`.

File: src/script/chem/molfile/molfile.js [hunk 6]
```diff
@@ -286,9 +286,135 @@ Molfile.prototype.writeCTab3000 = function (molecule) { // eslint-disable-line m
...
+			}
+			if (sgroup.data.queryOp.length > 0) {
+				// queryop is the query operator
+				sgDetails += ` QUERYOP=${sgroup.data.queryOp}`;
+			}
+			if (sgroup.data.fieldValue.length > 0) {
+				// fielddata is the query or field data.
+				sgDetails += ` FIELDDATA=${sgroup.data.fieldValue}`;
+			}
+		}
```
Comment (mycrobe) [inline src/script/chem/molfile/molfile.js]: could be extracted to reduce function size

**Comment by mycrobe** [comment]
Is there a test-case mol file that includes this functionality?

File: src/script/chem/molfile/molfile.js (hunk 2 above)
Comment (apar03) [reply src/script/chem/molfile/molfile.js]: Aah actually this did seem cryptic to me as well :(

I should have mentioned, slipped from my head somehow - I have plucked the "algorithm" to iterate over the S-Groups in the molecule from the V2000 writer, in this file itself -https://github.com/schrodinger/ketcher/blob/master/src/script/chem/molfile/molfile.js#L630

Agreed that the variable naming isn't that great and gives little information.
But I'll try my best renaming them here, with what best I can deduce

Comment (apar03) [reply src/script/chem/molfile/molfile.js]: sgforest#getSGroupBFS returns an array of s-group ids in the order of breadth-first search

File: src/script/chem/molfile/molfile.js (hunk 5 above)
Comment (apar03) [reply src/script/chem/molfile/molfile.js:311]: +1!

@pradeepnschrodinger  and I [discussed this offline](https://github.com/schrodinger/ketcher/pull/13#discussion_r386988111) to have a generic ticket to create utils for such purposes, and possibly perform some clean-up/refactor.

**Comment by apar03** [comment]
> Is there a test-case mol file that includes this functionality?

Some updation would be required in the [molcompare utility](https://github.com/schrodinger/ketcher/blob/master/test/utils/molcompare.js) for comparing representations (the static one and the one generated by Ketcher), specifically for S-Groups.

Will spin up a ticket for this, so that we can include a test molecule which specifically targets S-Groups in the future. [Edit - ticket [here](https://jira.schrodinger.com/browse/SS-29840)]

For testing on my end, I tried out by importing S-Groups via third party sketcher (Marvin), and then exporting these molecules in V3000 format in Ketcher. Then finally importing this representation back to Marvin, to verify that the molecule is exactly what we had initially sketched.

File: src/script/chem/molfile/molfile.js (hunk 3 above)
Comment (apar03) [reply src/script/chem/molfile/molfile.js]: Done

File: src/script/chem/molfile/molfile.js (hunk 6 above)
Comment (apar03) [reply src/script/chem/molfile/molfile.js]: Done

File: src/script/chem/molfile/molfile.js (hunk 5 above)
Comment (apar03) [reply src/script/chem/molfile/molfile.js:311]: ticket here https://jira.schrodinger.com/browse/SS-29875

**Review by mycrobe — APPROVED** [review]

**Pull request merged**